File: Backtracking/46.permutations.go
package leetcode

// @leet start
func permute(nums []int) [][]int {
	n := len(nums)
	res, path, seen := make([][]int, 0), make([]int, n), make([]bool, n)

	var dfs func(int)

	dfs = func(idx int) {
		if idx == n {
			res = append(res, append([]int{}, path...))
			return
		}
		// for i, v := range nums{ // this i will shaow the global i
		for i, v := range nums { //
			if seen[i] {
				continue
			}
			// path[i] = v need to change idx, to i
			path[idx] = v
			seen[i] = true
			dfs(idx + 1)
			seen[i] = false
		}
	}
	dfs(0)
	return res
}

// @leet end

File: Backtracking/46.permutations.ts
// @leet start
function permute(nums: number[]): number[][] {
  const res: number[][] = [];
  const path: number[] = [];
  const seen: boolean[] = nums.map(() => false);
  function dfs(i: number) {
    if (i === nums.length) {
      res.push(path.slice());
      return;
    }
    for (const [idx, v] of nums.entries()) {
      if (seen[idx]) {
        continue;
      }
      seen[idx] = true;
      path[i] = v;
      dfs(i + 1);
      seen[idx] = false;
    }
  }
  dfs(0);
  return res;
}
// @leet end

